### PR TITLE
Feature/support bnodes

### PR DESCRIPTION
**Pull request by Matthew-Kulich**
Merging this PR to csv-no-header branch, because after merge I will have to adjust tests to correspond to now feature.

### CODE REVIEW

**Review by blcham — APPROVED** [review]

**Review by blcham — CHANGES_REQUESTED** [review]
Not reviewed yet.

File: s-pipes-modules/module-tabular/src/main/java/cz/cvut/spipes/modules/util/BNodesTransformer.java [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+import org.apache.jena.util.ResourceUtils;
+
+import java.util.*;
+
+public class BNodesTransformer {
+
+    private BNodesTransformer() {}
+    private static Map<Resource, Resource> convertedNodes =  new HashMap<>();
+
+    public static Model convertBNodesToNonBNodes(Model model) {
```
Comment (blcham) [inline s-pipes-modules/module-tabular/src/main/java/cz/cvut/spipes/modules/util/BNodesTransformer.java:15]: ? static ?

IMHO, this won't work if the pipeline would have Tabular module called twice or if you execute the pipeline twice in parallel. What do u think?

I guess you need to call something like:
b = new BNodesTransormer;
b.convertBNodesToNonBNodes()

File: s-pipes-modules/module-tabular/src/test/resources/examples/lessColumns/input-data-schema.ttl [hunk 2]
```diff
@@ -14,15 +14,15 @@
 
 :bookingno_i_column   a csvw:Column;
     csvw:name "bookingno_i" ;
-    csvw:property :bookingno_i ;
+    csvw:propertyUrl :bookingno_i ;
```
Comment (blcham) [inline s-pipes-modules/module-tabular/src/test/resources/examples/lessColumns/input-data-schema.ttl:17]: IMO, this is not correct according to specification. If I remember correctly range of propertyUrl is Literal.

On the other hand, we wanted to define something like `csvw-e:property` that would link to Resource,

Comment (blcham) [reply s-pipes-modules/module-tabular/src/test/resources/examples/lessColumns/input-data-schema.ttl:17]: But we already discussed it  ...

Here are some fragments from documentation:

```
uriTemplate dfs:subClassOf xsd:string
```
```
propertyUrl
csvw:propertyUrl with string values interpreted as csvw:uriTemplate
````

File: s-pipes-modules/module-tabular/src/test/resources/examples/withProperty/expected-output.ttl [hunk 3]
```diff
@@ -22,14 +24,14 @@ _:b0    <http://www.w3.org/ns/csvw#name>
...
-                              1 ;
+                             "1"^^<http://www.w3.org/2001/XMLSchema#int> ;
                       <http://www.w3.org/ns/csvw#url>
                               <http://onto.fel.cvut.cz/data/#row=2>
                     ] ;
             <http://www.w3.org/ns/csvw#tableSchema>
                     [ a       <http://www.w3.org/ns/csvw#TableSchema> ;
                       <http://www.w3.org/ns/csvw#aboutUrl>
-                              "http://onto.fel.cvut.cz/data/#row-{_row}"^^<http://www.w3.org/ns/csvw#uriTemplate> ;
+                              "http://onto.fel.cvut.cz/data/#row-{_row}" ;
```
Comment (blcham) [inline s-pipes-modules/module-tabular/src/test/resources/examples/withProperty/expected-output.ttl:34]: Don't understand why we removed this.

File: s-pipes-modules/module-tabular/src/test/resources/examples/withProperty/expected-output.ttl [hunk 4]
```diff
@@ -46,17 +48,19 @@ _:b0    <http://www.w3.org/ns/csvw#name>
           ]
 ] .
 
-_:b4    <http://www.w3.org/ns/csvw#name>
+_:b4    a       <http://www.w3.org/ns/csvw#Column> ;
+        <http://www.w3.org/ns/csvw#name>
                 "user_sign" ;
         <http://www.w3.org/ns/csvw#propertyUrl>
-                "http://onto.fel.cvut.cz/data/user_sign_p"^^<http://www.w3.org/ns/csvw#uriTemplate>  ;
+                <http://onto.fel.cvut.cz/data/user_sign_p>  ;
```
Comment (blcham) [inline s-pipes-modules/module-tabular/src/test/resources/examples/withProperty/expected-output.ttl:55]: see comment above?

File: s-pipes-modules/module-tabular/src/test/resources/examples/withProperty/expected-output.ttl [hunk 5]
```diff
@@ -74,10 +78,12 @@ _:b2    <http://www.w3.org/ns/csvw#name>
         <http://onto.fel.cvut.cz/data/user_sign_p>
                 "KULICMAT" .
 
-_:b3    <http://www.w3.org/ns/csvw#name>
+_:b3    a       <http://www.w3.org/ns/csvw#Column> ;
+        <http://www.w3.org/ns/csvw#name>
                 "Start_Date" ;
         <http://www.w3.org/ns/csvw#propertyUrl>
-                "http://onto.fel.cvut.cz/data/Start_Date_p"^^<http://www.w3.org/ns/csvw#uriTemplate>  ;
+                <http://onto.fel.cvut.cz/data/Start_Date_p>  ;
```
Comment (blcham) [inline s-pipes-modules/module-tabular/src/test/resources/examples/withProperty/expected-output.ttl:85]: there are many issues like this one .... this is not correct

**Review by blcham — CHANGES_REQUESTED** [review]

**Comment by blcham** [comment]
csvw:propertyUrl -- semantika je definovana, nesmime predefinovat, xsd:range je xsd:string. Kdykoli generujeme (na vystupe) schemu musi byt csvw:propertyUrl pouzita. Na vstupe muze byt i jenom "csvw-e:property".
csvw-e:property -- todle si definujem sami a xsd:range je rdf:Resource

**Review by blcham — APPROVED** [review]
You got me here!

File: s-pipes-modules/module-tabular/src/test/resources/examples/withProperty/expected-output.ttl (hunk 3 above)
Comment (Matthew-Kulich) [reply s-pipes-modules/module-tabular/src/test/resources/examples/withProperty/expected-output.ttl]: Implemented

File: s-pipes-modules/module-tabular/src/test/resources/examples/withProperty/expected-output.ttl (hunk 5 above)
Comment (Matthew-Kulich) [reply s-pipes-modules/module-tabular/src/test/resources/examples/withProperty/expected-output.ttl]: Done

File: s-pipes-modules/module-tabular/src/test/resources/countries-model-output.ttl [hunk 6]
```diff
@@ -24,14 +24,14 @@
...
-                              1 ;
+                               "1"^^<http://www.w3.org/2001/XMLSchema#int> ;
                       <http://www.w3.org/ns/csvw#url>
                               <http://test-file#row=2>
                     ] ;
             <http://www.w3.org/ns/csvw#tableSchema>
                     [ a       <http://www.w3.org/ns/csvw#TableSchema> ;
                       <http://www.w3.org/ns/csvw#aboutUrl>
-                              "http://test-file#row-{_row}"^^<http://www.w3.org/ns/csvw#uriTemplate> ;
+                              "http://test-file#row-{_row}" ;
```
[the remark this answers is not in the document]
Comment (Matthew-Kulich) [reply s-pipes-modules/module-tabular/src/test/resources/countries-model-output.ttl]: Ok, the used version is 0.18.2

**Review by blcham — CHANGES_REQUESTED** [review]
LGTM, can be merged but after tests are not failing.

File: s-pipes-modules/module-tabular/src/test/java/cz/cvut/spipes/modules/TabularModuleTest.java [hunk 7]
```diff
@@ -163,7 +168,7 @@ void executeSelfWithNoDataSchemaNoHeaderReturnsAutonamedColumns()
 
         ExecutionContext outputContext = module.executeSelf();
 
-        for (int i = 1; i <= 3; i++) {
+        for (int i = 2; i <= 4; i++) {
```
[the remark this answers is not in the document]
Comment (Matthew-Kulich) [reply s-pipes-modules/module-tabular/src/test/java/cz/cvut/spipes/modules/TabularModuleTest.java:171]: It is the number of rows. The first one looks like `<http://onto.fel.cvut.cz/data/#row-2>`

File: s-pipes-modules/module-tabular/src/test/java/cz/cvut/spipes/modules/TabularModuleTest.java [hunk 8]
```diff
@@ -174,6 +179,34 @@ void executeSelfWithNoDataSchemaNoHeaderReturnsAutonamedColumns()
...
+         ExecutionContext outputContext = module.executeSelf();
+         Model actualModel = outputContext.getDefaultModel();
+
+         Model expectedModel = ModelFactory.createDefaultModel()
+                 .read(getFilePath("examples/blankNodes/expected-output.ttl").toString());
+
+         assertIsomorphic(actualModel, expectedModel);
+     }
+
+    void assertIsomorphic(Model actualModel, Model expectedModel){
```
[the remark this answers is not in the document]
Comment (Matthew-Kulich) [reply s-pipes-modules/module-tabular/src/test/java/cz/cvut/spipes/modules/TabularModuleTest.java:200]: Ok

**Comment by Matthew-Kulich** [comment]
@blcham I found the issue, look at the last commit, there was a bad prefix for `kbss-csvw`

**Review by blcham — APPROVED** [review]
LGTM

**Pull request merged**